Perf boost

JobLeonard opened the pull request:
So I just spent entirely too much time this evening/night trying to make `left-pad` faster, as can be seen by [the number of benchmarks I've created on perf.zone](https://perf.zone/user/tlh545IWtPgS86I7SAF8Qjma6w32):
![leftpad-2](https://user-images.githubusercontent.com/259840/36764219-dac48b8a-1c2b-11e8-85f3-bba14d21e2f3.gif)

After various dead ends (alternate ways of breaking out of the loop, loop unrolling), I finally found a set of tweaks that leads to consistent improvements for longer padding, _without_ reduction in performance when padding only a handful of characters.

There are a few small tweaks that had consistent improvements, like hoisting the padding variable, and switching the order in which `ch` is added to `pad`, because [it turns out](https://run.perf.zone/view/String-concat-does-the-order-matter-1519755302670) to matter [a tiny bit](https://run.perf.zone/view/String-concat-does-the-order-matter-1-190-1519755398884). Furthermore, instead of *breaking out* of the while-loop when `len` is zero, we *continue* it early as long as `len` is greater than zero.

However, the biggest change is very clearly this:

```js
    // ==== OLD ====
    if (ch === ' ' && len < 10) return cache[len] + str;
    var pad = '';
   
    // ==== NEW ====
    if (ch === ' ') {
        if (len < 16) return cache[len] + str;
        pad = cache[len & 15];
        len >>= 4;
        ch = '                ';
    }
```

Expanding the cache let's use skip the while loop when padding 10-15 characters. That's just a side-effect of the code below it, which leads to enormous improvements when padding large numbers of characters: `pad = cache[len & 15];  len >>= 4; ch = '                ';`
This initiates `pad` at the right size to let us skip the first four iterations of the loop. Recall that padding by 256 characters adds up to eight iterations. So this halved the time it takes to do that (minus a little bit of overhead).

Note that this would also work with `pad = cache[len & 7];  len >>= 3; ch = '        ';`, and in fact the benchmarks include that version. As you can see below, the 16-element version gives a much, _much_ higher benefit, and IMO isn't so much oversized as the next step (32 elements) would be.

# Selected, probably-bug-free benchmarks

## neither regression nor improvement
- [pad none](https://run.perf.zone/view/pad-none-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519783877604), 
- [pad 4](https://run.perf.zone/view/pad-4-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519781681326) (run multiple times to see how the "winner" varies each run), 
- [pad 8](https://run.perf.zone/view/pad-8-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519781475410)

## Significant perf boosts

- [pad 12](https://run.perf.zone/view/pad-12-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519781278596) (kind of"cheating", since we bumped the cache size)
- [pad 25](https://run.perf.zone/view/pad-25-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519781088832)
- [pad 50](https://run.perf.zone/view/pad-50-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519780882720)
- [pad 90](https://run.perf.zone/view/pad-90-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519780543048)
- [pad 9001](https://run.perf.zone/view/pad-9001-leftpad-bugfix-Steve-Maos-baseline-more-variants-1519780208772)

# Benchmark measurement gallery

## Chrome (Windows 10 64 bits)
![image](https://user-images.githubusercontent.com/259840/36765268-a4bca89c-1c30-11e8-8f93-f8f91b8e3661.png)

![image](https://user-images.githubusercontent.com/259840/36765275-b0951302-1c30-11e8-8904-e786aa8c0976.png)

![image](https://user-images.githubusercontent.com/259840/36765279-b6a435f2-1c30-11e8-9fb1-73135be472fe.png)

![image](https://user-images.githubusercontent.com/259840/36765283-bcc070c2-1c30-11e8-837e-40150468928d.png)

![image](https://user-images.githubusercontent.com/259840/36766080-8e527768-1c34-11e8-9bd6-10ee986335b7.png)

## Firefox 58 (Windows 10, 64 bits)

![image](https://user-images.githubusercontent.com/259840/36765349-05266a6a-1c31-11e8-8021-25c463ef4bbc.png)

![image](https://user-images.githubusercontent.com/259840/36765343-fd5896dc-1c30-11e8-84df-27cd7b63cb0a.png)

![image](https://user-images.githubusercontent.com/259840/36765333-f4eca0b0-1c30-11e8-87a3-15d529e10754.png)

![image](https://user-images.githubusercontent.com/259840/36765326-ec5db3da-1c30-11e8-9b0e-ca72ed24c2e7.png)

![image](https://user-images.githubusercontent.com/259840/36765899-c735badc-1c33-11e8-8376-d87350649cb9.png)

JobLeonard committed:
anywhere between 2x to >70x (!) perf boost

JobLeonard committed:
generate cache, bump it to 32

JobLeonard commented:
Made cache generated - it's less characters, and caching up to length 32 adds up to about a kilobyte of memoised strings. Probably not a significant overhead, but a hu gain in perf in many circumstances.

The following benchmark also include a 256-sized cache, which would be around 64 KiB of memoisation overhead. Given that it's not a very common use-case, I stuck to the maximum length of 32.

[pad 20](https://run.perf.zone/view/pad-20-leftpad-humongous-cache-1519791225526)
[pad 10](https://run.perf.zone/view/pad-10-leftpad-humongous-cache-1519791398388)
[pad 5](https://run.perf.zone/view/pad-5leftpad-humongous-cache-1519791633330)
[pad 500](https://run.perf.zone/view/pad-500-leftpad-humongous-cache-1519791827316)

stevemao commented:
CC @zehua @haoel @benjamingr @cht8687

zehua commented:
can you also try the perf tests in the perf folder? and test in node.js?

benjamingr commented on index.js:
```diff
@@ -6,46 +6,55 @@
+      // each time `ch` is "doubled", the `len` would need to be "doubled" too
+      // similar to finding a value in binary search tree, hence O(log(n))
+      ch += ch;
+      // only reached is `len` is not zero, so we can restart loop early
```
"if"

JobLeonard replied:
I don't see what you mean? You think there is a missing if? It's right there at line 48

benjamingr replied:
I meant the text, "only reached **is** `len` is not zero"

JobLeonard replied:
Ahaha! Good catch. Sorry, I wrote this around... like... 5 AM after a night of hyperfocused trial-and-error-programming and forgetting the need to sleep (so this definitely needs some more bug-testing too)

benjamingr replied:
Well, thanks for picking this up - if you're interested in putting "debug, optimize, repeat" skills into use - we could also use an extra hand at Node improving stuff :)

JobLeonard replied:
I'll be looking for part-time freelancing work starting in May, actually

benjamingr replied:
I can't pay you for Node (I don't get paid either) but the problems are really interesting https://github.com/nodejs/node/blob/master/CONTRIBUTING.md :)

JobLeonard replied:
Ah, like that! Well, if I have time :) 
So many fun projects out there that I would love to tinker with...

benjamingr commented:
So, I don't really trust these benchmarks :D 

Would you be able to run them using the Node benchmarking tool so we can see statistical significance? Alternatively, it would be nice to see an asm dump from v8.

benjamingr commented:
That said, to clarify, I believe this is a speed improvement - just not digging the benchmarks :D

benjamingr commented on index.js:
```diff
@@ -6,46 +6,55 @@
-  '        ',
-  '         '
-];
+// Generate cache. Strings are UTF16 in JavaScript,
```
So, the length 10 was picked arbitrarily after looking at the most common cases. It helps significantly in very common cases (aligning with spaces), but I don't feel too strongly about it.

Also, note the engine has no obligation to store these strings as 528*2 bytes.

JobLeonard replied:
You're completely right; I recall reading a mozilla blog post about the switch to faster, leaner data structures (some kind of rope, presumably). Speculating: in that case generating, the cached strings might also ensure better use of those internal data structures, making it a good idea regardless of what size it becomes.

Still, when in doubt I prefer to assume worst-case; like for people stuck with older browsers

benjamingr replied:
I'll clarify, the whole "optimization" of left pad that was done was for Node (where it might run a lot in build systems and servers) is _specifically_ based on the fact V8 represents strings with trees and collapses them to character arrays when it needs to - and the construction of the tree collapses anyway.

You can `padStart` and `padEnd` with modern versions anyway - I wanted to see if collapsing improved it and by how much :D

JobLeonard replied:
> I'll clarify, the whole "optimization" of left pad that was done was for Node (where it might run a lot in build systems and servers) 

Aaah, ok. 

A bit of background info: I have been programming for decades, but I only started looking at JS in the last two years, and even then in complete isolation on a very specific niche SPA for biologists. Since they use python on the backend, because SciPy, I have not used Node at all in that time (`npm` doesn't really count, I think). So I have huge gaps in contextual understanding of how the code is used in practice.

If you have any more feedback on what kind of optimisation matters, let me know.

JobLeonard commented:
@zehua: I had a suspicion that the bit patterns may have an influence on performance, so I even added a bunch of extra tests for that:

```
10922 (0b10101010101010)
 8193 (0b10000000000001)
 8192 (0b10000000000000)
 8191  (0b1111111111111)
 5461  (0b1010101010101)
 4096  (0b1000000000000)
 4095   (0b111111111111)
 2079   (0b100000011111)
 2048   (0b100000000000)
 2047    (0b11111111111)
 1365    (0b10101010101)
 1025    (0b10000000001)
 1024    (0b10000000000)
 1023     (0b1111111111)
```

It turns out the bit patterns do match: the more bits are set, the slower the algorithms are (except for `O(n)`). Makes sense, as these cases require more string concatenation after all.

I also tried a whole bunch of alternative tweaks, but none of them worked out:

```js
  // New while-loop
  while (true) {
    if (len & 1) pad = ch + pad;
    if (len >>= 1) {
      ch += ch;
      continue;
    }
    break;
  }

  // Switch, two bits. Three bits is similar, but with cases 7 to 0
  while (true) {
    ch2 = ch + ch;
    switch(len&3){
      case 3: //
        pad = ch2 + ch + pad;
        break;
      case 2:
        pad = ch2 + pad;
        break;
      case 1:
        pad = ch + pad;
      case 0:
      default:
    }
    if (len >>= 2) {
      ch = ch2 + ch2;
      continue;
    }
    break;
  }

  // Array lookup. The idea being that maybe branching was the bottleneck
  // var c = ['', '', '', ''];
  while (true) {
    c[2] = c[1] + c[1];
    pad = c[len&2] + (c[len&1] + pad);
    if (len >>= 2) {
      c[1] = c[2] + c[2];
      continue;
    }
    break;
  }
``` 


Anyway, the output of running all of the performance tests, after filtering the discarded alternative implementations:

```
-> pad 100 spaces to str of len 4
O(n)         x 1,578,867 ops/sec ±1.27% (90 runs sampled)
ES6 Repeat   x 8,546,499 ops/sec ±0.86% (89 runs sampled)
Current      x 8,829,279 ops/sec ±0.85% (90 runs sampled)
New          x 12,440,392 ops/sec ±0.57% (92 runs sampled)
-> pad  33 spaces to str of len 4
O(n)         x 3,493,611 ops/sec ±0.51% (92 runs sampled)
ES6 Repeat   x 8,907,255 ops/sec ±0.62% (95 runs sampled)
Current      x 9,583,444 ops/sec ±0.45% (94 runs sampled)
New          x 19,276,613 ops/sec ±0.63% (93 runs sampled)
-> pad  32 spaces to str of len 4
O(n)         x 3,549,147 ops/sec ±0.64% (90 runs sampled)
ES6 Repeat   x 9,081,341 ops/sec ±0.66% (93 runs sampled)
Current      x 9,712,771 ops/sec ±0.57% (93 runs sampled)
New          x 27,114,519 ops/sec ±0.62% (92 runs sampled)
-> pad  31 spaces to str of len 4
O(n)         x 3,536,994 ops/sec ±1.06% (89 runs sampled)
ES6 Repeat   x 7,672,999 ops/sec ±0.59% (89 runs sampled)
Current      x 8,131,713 ops/sec ±0.81% (92 runs sampled)
New          x 25,577,231 ops/sec ±1.56% (88 runs sampled)
-> pad  10 spaces to str of len 4
O(n)         x 5,901,201 ops/sec ±0.74% (95 runs sampled)
ES6 Repeat   x 9,047,213 ops/sec ±0.53% (95 runs sampled)
Current      x 9,625,336 ops/sec ±0.72% (92 runs sampled)
New          x 24,329,822 ops/sec ±0.89% (87 runs sampled)
-> pad   9 spaces to str of len 4
O(n)         x 6,201,515 ops/sec ±0.52% (94 runs sampled)
ES6 Repeat   x 8,912,250 ops/sec ±0.95% (93 runs sampled)
Current      x 22,530,869 ops/sec ±1.02% (89 runs sampled)
New          x 23,041,314 ops/sec ±0.77% (91 runs sampled)
-> pad  no spaces to str of len 4
O(n)         x 29,504,068 ops/sec ±0.88% (88 runs sampled)
ES6 Repeat   x 30,030,613 ops/sec ±1.15% (90 runs sampled)
Current      x 30,261,165 ops/sec ±0.96% (86 runs sampled)
New          x 29,898,400 ops/sec ±1.00% (90 runs sampled)


-> pad 100 spaces to str of len 100
O(n)         x 1,727,331 ops/sec ±0.75% (95 runs sampled)
ES6 Repeat   x 8,180,206 ops/sec ±0.81% (93 runs sampled)
Current      x 8,596,671 ops/sec ±0.59% (94 runs sampled)
New          x 11,964,140 ops/sec ±0.73% (92 runs sampled)
-> pad  10 spaces to str of len 100
O(n)         x 9,257,534 ops/sec ±0.60% (92 runs sampled)
ES6 Repeat   x 8,906,042 ops/sec ±0.75% (93 runs sampled)
Current      x 9,475,165 ops/sec ±0.67% (90 runs sampled)
New          x 21,960,796 ops/sec ±0.82% (87 runs sampled)
-> pad   9 spaces to str of len 100
O(n)         x 9,579,826 ops/sec ±0.88% (92 runs sampled)
ES6 Repeat   x 8,868,416 ops/sec ±0.65% (92 runs sampled)
Current      x 23,086,294 ops/sec ±0.91% (91 runs sampled)
New          x 22,893,180 ops/sec ±0.78% (92 runs sampled)
-> pad   0 spaces to str of len 100
O(n)         x 15,015,491 ops/sec ±0.91% (90 runs sampled)
ES6 Repeat   x 12,319,054 ops/sec ±0.68% (86 runs sampled)
Current      x 23,131,907 ops/sec ±0.89% (90 runs sampled)
New          x 22,928,214 ops/sec ±0.75% (88 runs sampled)


-> pad 10922 (0b10101010101010) spaces to str of len 4
O(n)         x 141,780 ops/sec ±0.94% (92 runs sampled)
ES6 Repeat   x 6,245,441 ops/sec ±0.71% (88 runs sampled)
Current      x 6,520,407 ops/sec ±0.59% (93 runs sampled)
New          x 7,412,821 ops/sec ±0.73% (92 runs sampled)
-> pad  8193 (0b10000000000001) spaces to str of len 4
O(n)         x 20,789 ops/sec ±1.04% (88 runs sampled)
ES6 Repeat   x 6,503,653 ops/sec ±0.71% (92 runs sampled)
Current      x 6,663,526 ops/sec ±0.61% (90 runs sampled)
New          x 7,079,254 ops/sec ±0.70% (92 runs sampled)
-> pad  8192 (0b10000000000000) spaces to str of len 4
O(n)         x 20,953 ops/sec ±1.03% (89 runs sampled)
ES6 Repeat   x 6,718,228 ops/sec ±0.70% (91 runs sampled)
Current      x 6,955,801 ops/sec ±0.82% (92 runs sampled)
New          x 7,116,594 ops/sec ±0.68% (92 runs sampled)
-> pad  8191  (0b1111111111111) spaces to str of len 4
O(n)         x 20,928 ops/sec ±0.97% (91 runs sampled)
ES6 Repeat   x 4,664,857 ops/sec ±0.94% (94 runs sampled)
Current      x 4,834,961 ops/sec ±0.63% (93 runs sampled)
New          x 5,155,518 ops/sec ±0.65% (91 runs sampled)


-> pad  5461  (0b1010101010101) spaces to str of len 4
O(n)         x 31,568 ops/sec ±0.87% (86 runs sampled)
ES6 Repeat   x 5,741,766 ops/sec ±0.80% (89 runs sampled)
Current      x 5,977,551 ops/sec ±0.54% (93 runs sampled)
New          x 6,505,239 ops/sec ±0.68% (92 runs sampled)
-> pad  4096  (0b1000000000000) spaces to str of len 4
O(n)         x 42,545 ops/sec ±1.03% (90 runs sampled)
ES6 Repeat   x 6,945,246 ops/sec ±0.60% (90 runs sampled)
Current      x 7,264,366 ops/sec ±0.82% (94 runs sampled)
New          x 7,616,480 ops/sec ±0.47% (93 runs sampled)
-> pad  4095   (0b111111111111) spaces to str of len 4
O(n)         x 42,363 ops/sec ±1.06% (90 runs sampled)
ES6 Repeat   x 5,191,652 ops/sec ±0.82% (95 runs sampled)
Current      x 5,241,190 ops/sec ±1.54% (85 runs sampled)
New          x 5,544,186 ops/sec ±1.22% (93 runs sampled)


-> pad  2079   (0b100000011111) spaces to str of len 4
O(n)         x 92,721 ops/sec ±0.79% (90 runs sampled)
ES6 Repeat   x 5,952,327 ops/sec ±0.76% (87 runs sampled)
Current      x 6,189,434 ops/sec ±0.51% (94 runs sampled)
New          x 7,888,843 ops/sec ±0.71% (95 runs sampled)
-> pad  2048   (0b100000000000) spaces to str of len 4
O(n)         x 94,377 ops/sec ±0.67% (94 runs sampled)
ES6 Repeat   x 7,143,774 ops/sec ±0.57% (91 runs sampled)
Current      x 7,405,973 ops/sec ±0.79% (93 runs sampled)
New          x 8,260,003 ops/sec ±0.56% (93 runs sampled)
-> pad  2047    (0b11111111111) spaces to str of len 4
O(n)         x 93,449 ops/sec ±0.83% (93 runs sampled)
ES6 Repeat   x 5,468,892 ops/sec ±0.55% (94 runs sampled)
Current      x 5,630,172 ops/sec ±0.75% (88 runs sampled)
New          x 6,671,885 ops/sec ±0.57% (94 runs sampled)


-> pad  1365    (0b10101010101) spaces to str of len 4
O(n)         x 142,199 ops/sec ±0.71% (90 runs sampled)
ES6 Repeat   x 6,251,298 ops/sec ±0.75% (92 runs sampled)
Current      x 6,549,310 ops/sec ±0.52% (94 runs sampled)
New          x 7,673,830 ops/sec ±0.74% (93 runs sampled)
-> pad  1025    (0b10000000001) spaces to str of len 4
O(n)         x 188,464 ops/sec ±0.69% (95 runs sampled)
ES6 Repeat   x 7,169,215 ops/sec ±0.61% (89 runs sampled)
Current      x 7,510,027 ops/sec ±0.65% (95 runs sampled)
New          x 8,596,675 ops/sec ±0.62% (91 runs sampled)
-> pad  1024    (0b10000000000) spaces to str of len 4
O(n)         x 178,073 ops/sec ±1.60% (87 runs sampled)
ES6 Repeat   x 7,082,301 ops/sec ±1.42% (87 runs sampled)
Current      x 7,545,679 ops/sec ±2.29% (87 runs sampled)
New          x 8,483,300 ops/sec ±1.18% (87 runs sampled)
-> pad  1023     (0b1111111111) spaces to str of len 4
O(n)         x 189,300 ops/sec ±0.74% (91 runs sampled)
ES6 Repeat   x 5,775,324 ops/sec ±0.69% (91 runs sampled)
Current      x 5,836,955 ops/sec ±1.89% (90 runs sampled)
New          x 7,430,372 ops/sec ±0.77% (92 runs sampled)
```

[perf-2018-03-01.txt](https://github.com/stevemao/left-pad/files/1769563/perf-2018-03-01.txt)

JobLeonard committed:
More tweaks

JobLeonard committed:
Test padding by 100, 1000 and 10000

JobLeonard commented:
So I had the following realisation: if `len > 1`, then `len&1` will always be true on the last iteration through the while loop, since the last bit shifted out *must* be set. We can make use of that as follows:

```js
// Current
while (true) {
  if (len & 1) pad += ch;
  len >>= 1;
  if (len) ch += ch;
  else break;
}
return pad + str;
```
```js
// "Previous new version"
while (true) {
  if (len & 1) pad = ch + pad;
  if (len >>= 1) {
    ch += ch;
    continue;
  }
  break;
}
return pad + str;
```
```js
// "Last iteration is always one"-optimisation
while (len>1) {
  if (len & 1) pad = ch + pad;
  ch += ch;
  len >>= 1;
}
return ch + pad + str
```
[2018-03-01 while perf test.txt](https://github.com/stevemao/left-pad/files/1771500/2018-03-01.while.perf.test.txt)

It seems to improve things a little bit further in almost every situation, except (for some reason) padding with nothing

JobLeonard commented:
More testing also revealed that loading from cache caused slowdown, so that was reduced from three cache loads (`pad`, early return and `ch`) back to one.
```js
// Reduced usage of loading cache
// ("cache" renamed "c" because minifiers didn't touch it)
var c = [' '];
for (var i = 1; i < 32; i++) {
  c.push(c[i - 1] + ' ');
}

  // Inside leftpad
  var pad = '';
  if (ch === ' ') {
    pad = c[len-1 & 31];
    if (len <= 32) return pad + str;
    ch = '                                ';
    len >>= 5;
  }
```

Finally, I tried throwing the function through a minifier to see what that would do, compare performance between it, and bring back the tweaks that caused improvements (if any). For some odd reason, turning the first statements into expressions inside `if` made it a bit faster:
```js
  if (str += '', len -= str.length, len <= 0) return str;
```
Well, if it works it works...

The minifier also turned a few if-statements into shortcircuited boolean expressions, but this actually slowed things down consistently across all benchmarks.

JobLeonard committed:
Mixed up `ch` string coercion

JobLeonard commented:
I wanted to verify the behaviour across a more elaborate range than the old benchmarks: in theory this version should have a more gradual performance drop, rather than steep a cliff, because it always makes use of the cache. I also thought it was rather fishy to not test the common cases padding by a few characters.

So I benchmarked padding by:

- 600, 300, 80, 60, 40, to verify the `O(log2(n))` behaviour in practice,

- 33, 32 and 31 to check behaviour around the cache-treshold of the new algorithm, 

- 10 and 9, which shows the jump in behaviour around the cache-treshold of the current algorithm

- 3, 2, 1 and 0, to see if any strange behaviour happens when padding by small amounts.

Of course, the benchmark results are now a gigantic list, so I added them as MD tables (hope they come out correctly). Alternatively, check out these text files. One is sorted by pad-size, the other by algorithm:

[2018-03-01 final perf sorted by algorithm.txt](https://github.com/stevemao/left-pad/files/1773127/2018-03-01.final.perf.sorted.by.algorithm.txt)

[2018-03-01 final perf sorted by test.txt](https://github.com/stevemao/left-pad/files/1773128/2018-03-01.final.perf.sorted.by.test.txt)

## Adding N to string of 4

| pad to 4 | O(n)                                        | ES6 Repeat                                  | Current                                     | New                                         |
|----------|---------------------------------------------|---------------------------------------------|---------------------------------------------|---------------------------------------------|
| 600      |    318,234 ops/sec ±0.82% (87 runs sampled) |  7,167,225 ops/sec ±0.68% (93 runs sampled) |  7,595,931 ops/sec ±0.59% (91 runs sampled) | 11,256,889 ops/sec ±0.66% (91 runs sampled) |
| 300      |    616,928 ops/sec ±0.69% (95 runs sampled) |  6,974,912 ops/sec ±0.63% (93 runs sampled) |  7,362,335 ops/sec ±0.62% (91 runs sampled) | 12,211,608 ops/sec ±0.77% (89 runs sampled) |
| 100      | 1,614,700 ops/sec ±0.57%  (93 runs sampled) | 8,334,206 ops/sec ±0.68%  (93 runs sampled) |  9,027,458 ops/sec ±0.63% (94 runs sampled) | 17,159,187 ops/sec ±0.66% (91 runs sampled) |
| 80       | 1,914,178 ops/sec ±0.84%  (95 runs sampled) |  8,094,001 ops/sec ±1.91% (88 runs sampled) |  8,595,445 ops/sec ±1.14% (88 runs sampled) | 17,245,765 ops/sec ±2.50% (84 runs sampled) |
| 60       |  2,223,685 ops/sec ±2.10% (87 runs sampled) |  7,646,457 ops/sec ±0.55% (94 runs sampled) |  8,049,007 ops/sec ±1.25% (89 runs sampled) | 22,009,381 ops/sec ±0.71% (91 runs sampled) |
| 40       |  3,062,008 ops/sec ±1.33% (83 runs sampled) |  8,826,190 ops/sec ±0.58% (93 runs sampled) |  9,313,352 ops/sec ±1.32% (87 runs sampled) | 22,312,363 ops/sec ±0.83% (90 runs sampled) |
| 33       |  3,472,875 ops/sec ±0.64% (94 runs sampled) |  9,116,922 ops/sec ±0.56% (89 runs sampled) |  9,501,558 ops/sec ±0.48% (91 runs sampled) | 23,184,105 ops/sec ±0.82% (91 runs sampled) |
| 32       |  3,565,452 ops/sec ±0.49% (94 runs sampled) |  9,141,453 ops/sec ±0.78% (92 runs sampled) |  9,937,253 ops/sec ±0.59% (91 runs sampled) | 27,944,963 ops/sec ±0.70% (88 runs sampled) |
| 31       |  3,600,233 ops/sec ±0.47% (93 runs sampled) |  7,836,050 ops/sec ±0.74% (94 runs sampled) |  8,371,875 ops/sec ±0.59% (91 runs sampled) | 27,696,111 ops/sec ±0.79% (91 runs sampled) |
| 10       |  6,210,148 ops/sec ±0.53% (92 runs sampled) |  9,011,449 ops/sec ±0.78% (94 runs sampled) |  9,484,077 ops/sec ±0.65% (90 runs sampled) | 26,409,714 ops/sec ±1.52% (87 runs sampled) |
| 9        |  6,361,138 ops/sec ±0.54% (93 runs sampled) |  8,880,046 ops/sec ±1.22% (90 runs sampled) | 22,554,466 ops/sec ±0.96% (89 runs sampled) | 27,766,146 ops/sec ±1.06% (90 runs sampled) |
| 3        | 12,482,995 ops/sec ±1.58% (88 runs sampled) | 10,329,966 ops/sec ±1.05% (89 runs sampled) | 17,440,063 ops/sec ±1.24% (84 runs sampled) | 20,340,355 ops/sec ±1.29% (87 runs sampled) |
| 2        | 14,803,566 ops/sec ±1.26% (88 runs sampled) | 13,262,128 ops/sec ±1.20% (88 runs sampled) | 19,182,810 ops/sec ±1.08% (87 runs sampled) | 20,391,334 ops/sec ±0.83% (89 runs sampled) |
| 1        | 19,927,483 ops/sec ±1.41% (85 runs sampled) | 15,201,355 ops/sec ±1.03% (90 runs sampled) | 18,161,628 ops/sec ±0.95% (90 runs sampled) | 18,841,426 ops/sec ±1.16% (89 runs sampled) |
| 0        | 32,254,390 ops/sec ±1.58% (83 runs sampled) | 34,085,066 ops/sec ±0.96% (90 runs sampled) | 33,266,201 ops/sec ±1.25% (87 runs sampled) | 33,477,011 ops/sec ±1.14% (88 runs sampled) |

## Adding N to string of 100

| pad to 100 | O(n)                                        | ES6 Repeat                                  | Current                                     | New                                         |
|------------|---------------------------------------------|---------------------------------------------|---------------------------------------------|---------------------------------------------|
| 600        |    290,782 ops/sec ±2.45% (86 runs sampled) |  6,564,973 ops/sec ±2.29% (87 runs sampled) |  7,140,991 ops/sec ±1.52% (86 runs sampled) | 10,018,212 ops/sec ±1.84% (86 runs sampled) |
| 300        |    600,427 ops/sec ±1.82% (88 runs sampled) |  6,779,161 ops/sec ±1.28% (92 runs sampled) |  6,978,974 ops/sec ±1.02% (92 runs sampled) | 11,052,600 ops/sec ±1.75% (84 runs sampled) |
| 100        |  1,651,645 ops/sec ±1.68% (90 runs sampled) |  7,785,310 ops/sec ±1.29% (87 runs sampled) |  8,174,350 ops/sec ±1.79% (87 runs sampled) | 14,897,696 ops/sec ±0.61% (91 runs sampled) |
| 80         |  2,102,781 ops/sec ±0.86% (93 runs sampled) |  7,669,593 ops/sec ±2.15% (87 runs sampled) |  8,460,189 ops/sec ±1.05% (90 runs sampled) | 17,404,901 ops/sec ±1.09% (87 runs sampled) |
| 60         |  2,674,569 ops/sec ±0.72% (91 runs sampled) |  7,531,679 ops/sec ±0.98% (88 runs sampled) |  7,745,863 ops/sec ±1.48% (88 runs sampled) | 18,664,724 ops/sec ±1.70% (86 runs sampled) |
| 40         |  3,518,480 ops/sec ±1.70% (86 runs sampled) |  8,150,397 ops/sec ±1.41% (91 runs sampled) |  8,656,107 ops/sec ±1.81% (84 runs sampled) | 19,186,349 ops/sec ±1.35% (89 runs sampled) |
| 33         |  4,146,534 ops/sec ±1.21% (89 runs sampled) |  8,695,667 ops/sec ±1.24% (89 runs sampled) |  8,927,099 ops/sec ±1.56% (86 runs sampled) | 19,032,654 ops/sec ±1.16% (90 runs sampled) |
| 32         |  4,105,962 ops/sec ±2.40% (79 runs sampled) |  8,800,633 ops/sec ±1.47% (89 runs sampled) |  9,118,866 ops/sec ±1.27% (87 runs sampled) | 22,898,605 ops/sec ±1.17% (88 runs sampled) |
| 31         |  4,353,146 ops/sec ±0.87% (93 runs sampled) |  7,457,086 ops/sec ±0.83% (87 runs sampled) |  7,913,758 ops/sec ±1.08% (92 runs sampled) | 22,295,158 ops/sec ±1.25% (87 runs sampled) |
| 10         |  9,037,193 ops/sec ±0.64% (92 runs sampled) |  8,672,538 ops/sec ±0.84% (87 runs sampled) |  9,562,590 ops/sec ±0.83% (89 runs sampled) | 23,113,400 ops/sec ±1.34% (88 runs sampled) |
| 9          |  9,726,867 ops/sec ±1.01% (90 runs sampled) |  8,423,409 ops/sec ±1.45% (87 runs sampled) | 21,154,178 ops/sec ±0.95% (88 runs sampled) | 21,767,360 ops/sec ±1.19% (89 runs sampled) |
| 3          | 16,653,521 ops/sec ±1.60% (88 runs sampled) | 11,920,777 ops/sec ±1.76% (87 runs sampled) | 21,955,858 ops/sec ±1.12% (85 runs sampled) | 20,070,573 ops/sec ±1.07% (90 runs sampled) |
| 2          | 18,579,902 ops/sec ±1.68% (82 runs sampled) | 14,184,847 ops/sec ±1.26% (85 runs sampled) | 22,528,312 ops/sec ±0.96% (92 runs sampled) | 21,525,254 ops/sec ±1.75% (82 runs sampled) |
| 1          | 22,799,335 ops/sec ±1.73% (89 runs sampled) | 17,182,478 ops/sec ±1.10% (91 runs sampled) | 23,031,531 ops/sec ±1.19% (86 runs sampled) | 23,113,619 ops/sec ±1.03% (87 runs sampled) |
| 0          | 32,169,194 ops/sec ±1.43% (88 runs sampled) | 30,652,651 ops/sec ±1.88% (86 runs sampled) | 31,281,872 ops/sec ±2.05% (84 runs sampled) | 32,213,494 ops/sec ±1.93% (86 runs sampled) |

JobLeonard commented:
A quick check to see what happens if the cache is bumped to length 64, and it's clear that we've hit the point of diminishing returns:
```
-> pad 600 spaces to str of len 4
Current      x 7,999,525 ops/sec ±0.60% (91 runs sampled)
New          x 11,503,735 ops/sec ±0.67% (89 runs sampled)
Fastest is New
-> pad 300 spaces to str of len 4
Current      x 7,397,365 ops/sec ±0.59% (95 runs sampled)
New          x 14,132,356 ops/sec ±0.54% (91 runs sampled)
Fastest is New
-> pad 100 spaces to str of len 4
Current      x 8,670,395 ops/sec ±0.78% (89 runs sampled)
New          x 17,976,067 ops/sec ±1.72% (86 runs sampled)
Fastest is New
-> pad  80 spaces to str of len 4
Current      x 9,005,222 ops/sec ±0.97% (94 runs sampled)
New          x 20,717,243 ops/sec ±0.80% (91 runs sampled)
Fastest is New
-> pad  60 spaces to str of len 4
Current      x 8,233,470 ops/sec ±0.89% (93 runs sampled)
New          x 23,970,087 ops/sec ±0.58% (90 runs sampled)
```

(Unless you are a Firefox user who *really* needs fast way to pad by a lot of spaces very often, [because then it's a different story](https://user-images.githubusercontent.com/259840/36877107-da3ed83a-1db8-11e8-9910-8152d87397d9.png))

Also, have you ever wondered _"hmm, I wonder what is the fastest way to pad a string with 100000000 spaces in JavaScript?"_ No? Well, I checked anyway:

```
-> pad 100000000 spaces recursively to empty string
Current - no recursion                x 4,180,644 ops/sec ±1.00% (91 runs sampled)
Current - 10000x10000                 x 4,465,338 ops/sec ±0.79% (88 runs sampled)
Current - 100x100x100x100             x 4,208,518 ops/sec ±2.38% (89 runs sampled)
Current - 10x10x10x10x10x10x10x10     x 3,934,396 ops/sec ±0.67% (93 runs sampled)
Current - 10x10x10x10x10000           x 4,079,508 ops/sec ±0.89% (91 runs sampled)
Current - 10000x10x10x10x10           x 3,968,951 ops/sec ±1.80% (88 runs sampled)

New Version - no recursion            x 4,262,856 ops/sec ±2.74% (86 runs sampled)
New Version - 10000x10000             x 5,202,351 ops/sec ±1.12% (92 runs sampled)
New Version - 100x100x100x100         x 5,278,240 ops/sec ±0.59% (95 runs sampled)
New Version - 10x10x10x10x10x10x10x10 x 3,967,643 ops/sec ±0.72% (93 runs sampled)
New Version - 10x10x10x10x10000       x 5,067,430 ops/sec ±0.66% (93 runs sampled)
New Version - 10000x10x10x10x10       x 4,973,443 ops/sec ±0.75% (91 runs sampled)

Fastest is New Version - 100x100x100x100, New Version - 10000x10000
```

JobLeonard committed:
merge ch if-statements, add 0 cache

JobLeonard commented:
What we were doing:
- check value of `ch`, default to `' '`
- check if `ch` is `' '`, if so replace it with cache.

We can smush that into one if-statement. And while we're at it, let's throw #46 a bone and cache the also-common `0` case as well:

```js
// var c_ = [' '], c0 = ['0'];
// for (var i = 1; i < 32; i++) {
//   c_.push(c_[i - 1] + ' ');
//   c0.push(c0[i - 1] + '0');
// }

if (ch === ' ' || !ch && ch !== 0){
  // We use a cache to speed up the default case
  pad = c_[len-1 & 31];
  if (len <= 32) return pad + str;
  // `pad` effectively contains the result of
  // iterating the loop below five times,
  // so we can skip those.
  // `ch` is set to 32 spaces to accomodate this.
  ch = '                                ';
  len >>= 5;
} else if ((ch+= '') === '0'){
  // padding by zero is another really common use-case
  pad = c0[len-1 & 31];
  if (len <= 32) return pad + str;
  ch = '00000000000000000000000000000000';
  len >>= 5;
}
```

JobLeonard commented:
Eh... I don't understand why test 41 is failing? It works fine off-line.

JobLeonard committed:
Node 0.12 fix attempt

JobLeonard committed:
Expand test.js to cover more edge cases

- truthy/falsy values
- potentially weird type coercion interactions
- more zero-padding test cases

Also gave everything a more descriptive message

JobLeonard commented:
For the record, I'm still working on this but it has been on the back-burner. Also, when I try to benchmark it I notice that *the order in which I do the benchmarks affects the results*, so it's really hard to get reliable information.

perhaps V8 caches previously used strings or something, which would make it really hard to get a good idea of what the fasters approach is.

benjamingr commented:
@JobLeonard as suggested before - consider using the node benchmarking tool which checks for statistical significance.

JobLeonard commented:
I am, and I had found that changing the order changes results

benjamingr commented:
@JobLeonard and are the scores significant? (How many *s in the output)

JobLeonard commented:
Look, you should not trust benchmarks that use scores in this way anyway, they're the equivalent to dynamite plots, and the problems with dynamite plots are... well, a picture says more than a thousand words:

![image](https://user-images.githubusercontent.com/259840/37905277-a0af15ea-30fe-11e8-88f5-31b9bf835a63.png)

Raw data vs mean + std deviation.

(taken from http://biostat.mc.vanderbilt.edu/wiki/pub/Main/RafeDonahue/fscipdpfcbg_currentversion.pdf, which is an excellent book on how to present statistics, BTW). See also [datasaurus dozen](https://www.autodeskresearch.com/publications/samestats) and [Anscombe's Quartet](https://en.wikipedia.org/wiki/Anscombe%27s_quartet). 

TL;DR: I want to see the distribution, dammit!

Anyway, it's been a few weeks, but yes, they "were". The benchmarks just flat-out stated that one was better with small error margins, yet this can change depending on:

- order of test inputs (long to short strings or the other way around),
- what kind of test inputs (`' '`, `undefined`, `0`, `'0'` - I'm testing them all)
- how many inputs (just a few? Repeat with a ton of different lengths? Does it include *extreme* lengths like padding with 10000+ characters?
- the order of algorithms

(I didn't think the existing benchmarks covered enough varieties on inputs, missing out on different behaviours. I extended them a *lot*, but it also takes much longer to benchmark now. I'll probably refactor them into different perf scripts or something)

So to be clear: this change would be persistent between benchmark runs with the one particular ordering, but not when changing these orders.

benjamingr commented:
> Look, you should not trust benchmarks that use scores in this way anyway,

You really confused me here...

First, the Node.js benchmarks let you plot the results - fon scatter plots - see scatter.R and for the box plot you show above see compare.R - all these things you say about benchmarks are true - but were already taken them into account in Node's benchmarking tools.

It also runs a [t-test](https://en.wikipedia.org/wiki/Student%27s_t-test) to make sure the results are significant and then reports the [p value](https://en.wikipedia.org/wiki/P-value) - those "stars" that I trust just indicate that P value (p < 0.05 gets a *, p < 0.01 gets ** and p < 0.001 gets ***)

JobLeonard commented:
> First, the Node.js benchmarks let you plot the results - fon scatter plots - see scatter.R and for the box plot you show above see compare.R - all these things you say about benchmarks are true - but were already taken them into account in Node's benchmarking tools.

Ok, maybe I'm going nuts here but there is no `scatter.R` or `compare.R` in this repo, nor in the [`benchmark.js`](
https://benchmarkjs.com/docs) documentation, which is the only package referred to in the benchmarking scripts *in this repo*. On the node website there is nothing referring to 
"node benchmarking tools" in any of the docs. If I google for `node scatter.R` I find [this](https://github.com/pmq20/node-packer/blob/master/node/benchmark/scatter.R), which is an R-script in the node-packer repository. So now *I* am confused: since when does node use *R* for its stats? That's a completely different language.


> It also runs a t-test to make sure the results are significant and then reports the p value - those "stars" that I trust just indicate that P value (p < 0.05 gets a *, p < 0.01 gets ** and p < 0.001 gets ***)

Look, all I've been doing is run `node perf` in the `/test/perf` folder and seeing output like this:

```txt
-> pad 40 "undefined"-s  (undefined) to str of len 4
New          x 19,071,148 ops/sec ±1.26% (86 runs sampled)
Minified     x 17,299,597 ops/sec ±0.97% (88 runs sampled)
Current      x 8,455,192 ops/sec ±1.10% (88 runs sampled)
Fastest is New
```

There is no star. I don't see any of the things that you are talking about.

I don't see any way of capturing different output based on how the script it written in this repository, or what I can see from looking at the benchmark.js API.

benjamingr commented:
@JobLeonard I'll clarify, when I say the Node benchmarks I meant running this tool and not benchmark.js: https://github.com/nodejs/node/tree/master/benchmark

JobLeonard commented:
And *of course* this is not linked anywhere from the main Node.js website...

Thanks Benjamin, you're a lifesaver

benjamingr commented:
@JobLeonard this is an internal tool we use at Node, I recommended it since it's the only benchmarking tool for JS that I know and is reasonable - sorry for not being clear about it!
